File: src/truthound_dashboard/schemas/artifacts.py
"""Canonical artifact schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import Field

from .base import BaseSchema, IDMixin, ListResponseWrapper, TimestampMixin


class ArtifactGenerateRequest(BaseSchema):
    format: str = Field(default="html")
    theme: str | None = Field(default="professional")
    locale: str = Field(default="en")
    title: str | None = Field(default=None)
    include_samples: bool = Field(default=True)
    include_statistics: bool = Field(default=True)
    custom_metadata: dict[str, Any] | None = Field(default=None)


class DataDocsGenerateRequest(BaseSchema):
    theme: str | None = Field(default="professional")
    title: str | None = Field(default=None)


class ArtifactLocaleInfo(BaseSchema):
    code: str
    english_name: str
    native_name: str
    flag: str
    rtl: bool = False


class ArtifactCapabilitiesResponse(BaseSchema):
    formats: list[str] = Field(default_factory=list)
    themes: list[str] = Field(default_factory=list)
    locales: list[ArtifactLocaleInfo] = Field(default_factory=list)
    artifact_types: list[str] = Field(default_factory=list)


class ArtifactResponse(BaseSchema, IDMixin, TimestampMixin):
    workspace_id: str | None = None
    source_id: str | None = None
    validation_id: str | None = None
    artifact_type: str
    format: str
    status: str
    title: str
    description: str | None = None
    file_path: str | None = None
    external_url: str | None = None
    file_size: int | None = None
    content_hash: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    error_message: str | None = None
    generation_time_ms: float | None = None
    expires_at: datetime | None = None
    downloaded_count: int = 0
    last_downloaded_at: datetime | None = None
    locale: str = "en"
    theme: str | None = None
    source_name: str | None = None
    validation_status: str | None = None
    is_expired: bool = False
    download_url: str | None = None

    @classmethod
    def from_model(cls, model: Any) -> "ArtifactResponse":
        return cls(
            id=str(model.id),
            workspace_id=str(model.workspace_id) if model.workspace_id else None,
            source_id=str(model.source_id) if model.source_id else None,
            validation_id=str(model.validation_id) if model.validation_id else None,
            artifact_type=model.artifact_type,
            format=model.format,
            status=model.status,
            title=model.title,
            description=model.description,
            file_path=model.file_path,
            external_url=model.external_url,
            file_size=model.file_size,
            content_hash=model.content_hash,
            metadata=model.artifact_metadata or {},
            error_message=model.error_message,
            generation_time_ms=model.generation_time_ms,
            expires_at=model.expires_at,
            downloaded_count=model.downloaded_count,
            last_downloaded_at=model.last_downloaded_at,
            locale=model.locale,
            theme=model.theme,
            source_name=model.source.name if getattr(model, "source", None) else None,
            validation_status=model.validation.status if getattr(model, "validation", None) else None,
            is_expired=model.is_expired,
            created_at=model.created_at,
            updated_at=model.updated_at,
        )


class ArtifactListResponse(ListResponseWrapper[ArtifactResponse]):
    pass


class ArtifactStatistics(BaseSchema):
    total_artifacts: int = 0
    by_type: dict[str, int] = Field(default_factory=dict)
    by_status: dict[str, int] = Field(default_factory=dict)
    total_downloads: int = 0
    total_size_bytes: int = 0

File: frontend/src/api/modules/artifacts.ts
import { request, ApiError, getStoredSessionToken } from '../core'

export type ArtifactFormat = 'html' | 'csv' | 'json'
export type ArtifactTheme = 'light' | 'dark' | 'professional' | 'minimal' | 'high_contrast'
export type ArtifactLocale =
  | 'en' | 'ko' | 'ja' | 'zh' | 'de' | 'fr' | 'es' | 'pt'
  | 'it' | 'ru' | 'ar' | 'th' | 'vi' | 'id' | 'tr'
export type ArtifactStatus = 'pending' | 'generating' | 'completed' | 'failed' | 'expired'

const API_BASE = '/api/v1'

export interface ArtifactRecord {
  id: string
  workspace_id?: string
  source_id?: string
  source_name?: string
  validation_id?: string
  validation_status?: string
  artifact_type: string
  format: ArtifactFormat
  status: ArtifactStatus
  title: string
  description?: string
  file_path?: string
  external_url?: string
  file_size?: number
  content_hash?: string
  metadata: Record<string, unknown>
  error_message?: string
  generation_time_ms?: number
  expires_at?: string
  downloaded_count: number
  last_downloaded_at?: string
  locale: ArtifactLocale
  theme?: ArtifactTheme
  is_expired: boolean
  download_url?: string
  created_at: string
  updated_at: string
}

export interface ArtifactListResponse {
  data: ArtifactRecord[]
  total: number
  offset: number
  limit: number
}

export interface ArtifactStatistics {
  total_artifacts: number
  by_type: Record<string, number>
  by_status: Record<string, number>
  total_downloads: number
  total_size_bytes: number
}

export interface ArtifactLocaleInfo {
  code: ArtifactLocale
  english_name: string
  native_name: string
  flag: string
  rtl: boolean
}

export interface ArtifactCapabilities {
  formats: ArtifactFormat[]
  themes: ArtifactTheme[]
  locales: ArtifactLocaleInfo[]
  artifact_types: Array<'report' | 'datadocs'>
}

export interface ArtifactGenerateOptions {
  format?: ArtifactFormat
  theme?: ArtifactTheme
  locale?: ArtifactLocale
  title?: string
  include_samples?: boolean
  include_statistics?: boolean
}

export async function listArtifacts(params?: {
  workspace_id?: string
  saved_view_id?: string
  search?: string
  status?: ArtifactStatus
  source_id?: string
  validation_id?: string
  artifact_type?: string
  format?: ArtifactFormat
  include_expired?: boolean
  offset?: number
  limit?: number
}): Promise<ArtifactListResponse> {
  const limit = params?.limit ?? 20
  const offset = params?.offset ?? 0
  const page = Math.floor(offset / limit) + 1

  return request<ArtifactListResponse>('/artifacts', {
    params: {
      workspace_id: params?.workspace_id,
      saved_view_id: params?.saved_view_id,
      search: params?.search,
      status: params?.status,
      source_id: params?.source_id,
      validation_id: params?.validation_id,
      artifact_type: params?.artifact_type,
      format: params?.format,
      include_expired: params?.include_expired,
      page,
      page_size: limit,
      offset,
      limit,
    },
  })
}

export async function getArtifactStatistics(): Promise<ArtifactStatistics> {
  return request<ArtifactStatistics>('/artifacts/statistics')
}

export async function getArtifactCapabilities(): Promise<ArtifactCapabilities> {
  return request<ArtifactCapabilities>('/artifacts/capabilities')
}

export async function getArtifact(artifactId: string): Promise<ArtifactRecord> {
  return request<ArtifactRecord>(`/artifacts/${artifactId}`)
}

export async function deleteArtifact(artifactId: string): Promise<void> {
  await request<void>(`/artifacts/${artifactId}`, { method: 'DELETE' })
}

export async function cleanupExpiredArtifacts(): Promise<{ deleted: number }> {
  return request<{ deleted: number }>('/artifacts/cleanup', { method: 'POST' })
}

export async function downloadArtifact(artifactId: string): Promise<Blob> {
  const url = `${API_BASE}/artifacts/${artifactId}/download`
  const headers = new Headers()
  const sessionToken = getStoredSessionToken()
  if (sessionToken) {
    headers.set('X-Truthound-Session', sessionToken)
  }
  const response = await fetch(url, { headers })
  if (!response.ok) {
    throw new ApiError(response.status, response.statusText)
  }
  return response.blob()
}

export async function generateReportArtifact(
  validationId: string,
  options?: ArtifactGenerateOptions
): Promise<ArtifactRecord> {
  return request<ArtifactRecord>(`/artifacts/validations/${validationId}/report`, {
    method: 'POST',
    body: JSON.stringify(options || {}),
  })
}

export async function generateDataDocsArtifact(
  validationId: string,
  options?: { title?: string; theme?: ArtifactTheme }
): Promise<ArtifactRecord> {
  return request<ArtifactRecord>(`/artifacts/validations/${validationId}/datadocs`, {
    method: 'POST',
    body: JSON.stringify(options || {}),
  })
}
